s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala

```diff
@@ -642,53 +602,99 @@ import scala.collection.immutable
+          Flow[(Chunk, (MultipartUpload, Int))]
+            .map {
+              case (chunkedPayload, (uploadInfo, chunkIndex)) =>
+                //each of the payload requests are created
+                val partRequest =
+                  uploadPartRequest(uploadInfo, chunkIndex, chunkedPayload.data, chunkedPayload.size, headers)
+                (partRequest, (uploadInfo, chunkIndex))
+            }
+            .flatMapConcat { case (req, info) => Signer.signedRequest(req, signingKey).zip(Source.single(info)) }
+            .via(superPool[(MultipartUpload, Int)])
 
-        requestFlow
-          .via(superPool[(MultipartUpload, Int)])
+        SplitAfterSize(chunkSize, chunkBufferSize)(atLeastOneByteString)
+          .via(getChunkBuffer(chunkSize, chunkBufferSize)) //creates the chunks
+          .concatSubstreams
+          .zip(requestInfo)
+          // Allow requests that fail with transient errors to be retried, using the already buffered chunk.
+          .via(RetryFlow.withBackoff(200.millis, 10.seconds, 0.0, maxRetriesPerChunk, retriableFlow) {
+            case (chunkAndUploadInfo, (Success(r), _)) =>
+              // 5xx errors from S3 can be treated as transient.
+              if (r.status.intValue >= 500) Some(chunkAndUploadInfo) else None
+            case (chunkAndUploadInfo, (Failure(_), _)) =>
+              // Treat any exception as transient.
+              Some(chunkAndUploadInfo)
+          })
           .mapAsync(parallelism) {
-            case (Success(r), (upload, index)) =>
-              if (r.status.isFailure()) {
-                Unmarshal(r.entity).to[String].map { errorBody =>
-                  FailedUploadPart(
-                    upload,
-                    index,
-                    new RuntimeException(
-                      s"Upload part $index request failed. Response header: ($r), response body: ($errorBody)."
-                    )
-                  )
-                }
-              } else {
-                r.entity.discardBytes()
-                val etag = r.headers.find(_.lowercaseName() == "etag").map(_.value)
-                etag
-                  .map(t => Future.successful(SuccessfulUploadPart(upload, index, t)))
-                  .getOrElse(
-                    Future
-                      .successful(FailedUploadPart(upload, index, new RuntimeException(s"Cannot find etag in ${r}")))
-                  )
-              }
-
-            case (Failure(e), (upload, index)) => Future.successful(FailedUploadPart(upload, index, e))
+            case (response, (upload, index)) => handleChunkResponse(response, upload, index)
           }
       }
       .mapMaterializedValue(_ => NotUsed)
   }
 
+  private def handleChunkResponse(response: Try[HttpResponse], upload: MultipartUpload, index: Int)(
+      implicit mat: Materializer
+  ) = {
+
+    import mat.executionContext
+
+    response match {
+      case Success(r) =>
+        if (r.status.isFailure()) {
```

Comment:
[minor] I know you're just moving this code around, but adding the conditional to the pattern match would make this cleaner.

Reply:
Done.